test/test_transformer.cc

```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
+  auto psWrite = allOf(access(_A, _ii, _jj));
+  auto matchedWrites =
+    match(writes, psWrite);
+  //std::cout << "matched Writes " << matchedWrites.size() << std::endl;
+
+  // tmp[j][i] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the checks. this is because we do not link
+  // read and write at the moment. Placeholder are _not_
+  // reused between different calls to allOf. We can overcome
+  // this inspecting the placeholder for the write and the read.
+  // They should be equal. 
+ 
+  matched = matched && (matchedWrites.size() == 1);
+  matched = matched && (matchedReads.size() == 1);
+  EXPECT_TRUE(matched);
+
+  // check index for read and write are equal
+  matched = matched && 
+    matchedWrites[0][_ii].payload().inputDimPos_ == matchedReads[0][_i].payload().inputDimPos_;
+  matched = matched &&
+    matchedWrites[0][_jj].payload().inputDimPos_ == matchedReads[0][_j].payload().inputDimPos_;
+  EXPECT_TRUE(matched);
+
+  // D[i][j] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the test. We may want to apply the same check
+  // as before also for the accessed array.
+
+  // step 1. (not implemented, just assumed)
+  // we assume the indexes are already in
+  // the "good" order. (i,j,k).
+  // Line 1285 to 1291 in ScheduleOptimizer.cpp
+
+  // step 2. create macro-kernel
+  // line 1294 in ScheduleOptimizer.cpp
+  // For the micro and macro kernels we assume
+  // given values for the tile size.
+  using namespace builders;
+  // tiling - Mc=96, Nc=2048, Kc=512
+  // set tile values manually
+  int dimOutNum = isl_schedule_node_band_n_member(node.get());
+  std::vector<int> tileSizes(dimOutNum);
+  tileSizes[dimOutNum - 3] = /*96*/32;
+  tileSizes[dimOutNum - 2] = /*2048*/32;
+  tileSizes[dimOutNum - 1] = /*512*/32;
+  // tile node and get partial schedule 
+  isl::schedule_node tiledNode = tile(node, tileSizes);
+  auto tileSchedule = tiledNode.band_get_partial_schedule();
+  auto pointSchedule = tiledNode.child(0).band_get_partial_schedule();
+  // id for mark nodes.
+  std::string markerTile = "first level tiling - tile";
+  isl::id idTile = isl::id::alloc(node.get_ctx(), markerTile, nullptr);
+  std::string markerPoint = "first level tiling - point";
+  isl::id idPoint = isl::id::alloc(node.get_ctx(), markerPoint, nullptr);
+
+  // clang-format off
+  ScheduleNodeBuilder macroKernel = 
+    mark(idTile,
+      band(
+        [&node, &tileSchedule, &dimOutNum]() {
+          auto newPartialSchedule = permuteDims(tileSchedule, dimOutNum - 2, dimOutNum - 1);
+          newPartialSchedule = permuteDims(newPartialSchedule, dimOutNum - 3, dimOutNum - 1);
+          return newPartialSchedule;},
+        mark(idPoint,
+	  band(pointSchedule))));
+  // clang-format on
+
+  if(matched) {
+    node = rebuild(node, macroKernel);
+    root = node.root();
+    //std::cout << root.to_str() << std::endl;
+  }
+
+  // match micro-kernel
+  auto matcherMicroKernel = [&]() {
+    using namespace matchers;
+    // clang-format off
```

Comment:
there's no need to turn the formatter of if you don't have the tree formatting in can break :)